Document APIs for the System.Composition.Convention Namespace

tompratt-AQ opened the pull request:
# Title
Document APIs for the System.Composition.Convention Namespace

## Summary
Added documentation for "To be added." fields in all classes of the System.Composition.Convention Namespace:

- ConventionBuilder 
- ExportConventionBuilder 
- ImportConventionBuilder 
- ParameterImportConventionBuilder 
- PartConventionBuilder 
- PartConventionBuilderOfT 

**Note:** I will be adding two more files for this namespace in another commit:
The AttributedModelProvider base class, and the namespace file ns-System.Composition.Convention.xml

Fixes #2261 

## Suggested Reviewers
@rpetrusha

tompratt-AQ committed:
Document APIs for the System.Composition.Convention Namespace

tompratt-AQ committed:
Added missing closing tags.

rpetrusha commented on xml/System.Composition.Convention/ImportConventionBuilder.xml:
```diff
@@ -147,9 +147,9 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <summary>Configures the import to receive all exports of the contract.</summary>
+        <returns>An import builder that is configured to receive all exports of the contract, and allows for further import configuration.</returns>
```
no comma after contract

rpetrusha commented on xml/System.Composition.Convention/ParameterImportConventionBuilder.xml:
```diff
@@ -12,8 +12,8 @@
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Contains methods used when configuring a <see cref="PartConventionBuilder{T}"/>. This class is used only in expressions, instances of this type are never created.</summary>
```
Move "This class is used..." to the Remarks section. Also, a semicolon, not a comma, should be used.

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder.xml:
```diff
@@ -34,11 +34,11 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="name">The metadata name.</param>
+        <param name="getValueFromPartType">A function that calculates the metadata value based on the type.</param>
```
calculates --> determines

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder.xml:
```diff
@@ -197,10 +197,10 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="interfaceFilter">The filter for selecting the interfaces.</param>
+        <summary>Selects the interfaces on the part type that will be exported.</summary>
```
that will be exported --> to export

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder.xml:
```diff
@@ -221,11 +221,11 @@
-        <remarks>To be added.</remarks>
+        <param name="interfaceFilter">The filter for selecting the interfaces.</param>
+        <param name="exportConfiguration">The action to configure the selected interfaces.</param>
+        <summary>Selects the interfaces on the part type that will be exported.</summary>
```
that will be exported --> to export

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder.xml:
```diff
@@ -476,10 +476,10 @@
-        <summary>To be added.</summary>
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="constructorSelector">Filter that selects a single constructor.</param>
```
Filter --> A filter

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder.xml:
```diff
@@ -476,10 +476,10 @@
-        <returns>To be added.</returns>
-        <remarks>To be added.</remarks>
+        <param name="constructorSelector">Filter that selects a single constructor.</param>
+        <summary>Selects which of the available constructors will be used to instantiate the part.</summary>
```
which of the available constructors will be used --> which constructor is used

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder.xml:
```diff
@@ -500,11 +500,11 @@
-        <remarks>To be added.</remarks>
+        <param name="constructorSelector">A filter that selects a single constructor.</param>
+        <param name="importConfiguration">The action configuring the parameters of the selected constructor.</param>
+        <summary>Selects which of the available constructors will be used to instantiate the part.</summary>
```
which of the available constructors will be used --> which constructor is used

rpetrusha reviewed (changes requested):
The namespace also needs to be documented. The namespace also contains a type named AttributedModelProvider that needs to be documented.

tompratt-AQ commented:
@rpetrusha Thanks for the detailed review. I'm working on the namespace topic and the base class topic right now.

tompratt-AQ committed:
Added namespace and base topics and made changes per review.

tompratt-AQ commented:
@rpetrusha I added AttributedModelProvider.xml and ns-System.Composition.Convention.xml for you to review, and made the suggested changes from your last review. Thanks!

rpetrusha commented on xml/System.Composition.Convention/AttributedModelProvider.xml:
```diff
@@ -12,8 +12,8 @@
   <Docs>
-    <summary>To be added.</summary>
-    <remarks>To be added.</remarks>
+    <summary>Defines the core behavior of methods used to retrieve metadata attributes for a type, and provides a base for derived classes.</summary>
```
This can be a bit clearer: Serves as the base class that defines the core behavior of methods used to retrieve metadata attributes for a type.

rpetrusha commented on xml/System.Composition.Convention/AttributedModelProvider.xml:
```diff
@@ -28,8 +28,8 @@
       <Docs>
-        <summary>To be added.</summary>
-        <remarks>To be added.</remarks>
+        <summary>Called from constructors in derived classes to initialize the <see cref="AttributedModelProvider"/> class.</summary>
```
Generally, we just want to document protected constructors as, "Creates a new instance of the <whatever> class."
In the Remarks, you can add a comment that this is a protected constructor that can only be called from derived classes.

tompratt-AQ committed:
Addressed review comments for AttributedModelProvider.

tompratt-AQ committed:
Added exceptions for this namespace.

tompratt-AQ commented:
@rpetrusha Could you please review the Exceptions that I added in this last commit to docs/xml/System.Composition.Convention/PartConventionBuilder`1.xml.  
They are thrown by internal methods that are called by the public methods to which I added them. They are custom Argument exceptions for this class, the code for them is in:  https://github.com/dotnet/corefx/blob/master/src/System.Composition.Convention/src/System/Composition/ExceptionBuilder.cs
They are the only exceptions in the namespace.  Thanks!

tompratt-AQ commented:
@mairaw Does this PR need to be restarted? It passed all checks last Friday and is waiting for a second review of a couple items (see my previous comment), but as of yesterday it has been stuck on these new additional checks.

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder`1.xml:
```diff
@@ -66,6 +66,7 @@
         <summary>Selects a property on the part to export.</summary>
         <returns>A part builder allowing further configuration of the part.</returns>
         <remarks></remarks>
+        <exception cref="T:System.Argument"><paramref name="propertySelector" /> The expression must be <see cref="T:System.Linq.Expressions.MemberExpression"/> for accessing a property.</exception>
```
`<paramref>` is a general formatting tag for use in text. So the exception description should read, "The `<paramref name="propertySelector" />` expression must be a `<see cref="T:System.Linq.Expressions.MemberExpression"/>` for accessing a property." That's true of each of the uses of the paramref element here.

Also, in each case, "a" should precede the type name ("a `<see cref="T:System.Linq.Expressions.MemberExpression"/>`"). There are a few cases where you've done that, but it should be done in all cases.

rpetrusha commented on xml/System.Composition.Convention/PartConventionBuilder`1.xml:
```diff
@@ -272,6 +277,7 @@
         <summary>Selects which constructor is used to instantiate the part.</summary>
         <returns>A part builder allowing further configuration of the part.</returns>
         <remarks></remarks>
+        <exception cref="T:System.Argument"><paramref name="constructorSelector" /> The expression must be new.</exception>
```
Instead of "expression must be new", "expression must use the `new` operator"

rpetrusha reviewed (commented):
@tompratt-AQ, two comments, but they apply to all of the exception items added.

tompratt-AQ committed:
Made corrections per review comments.

tompratt-AQ commented:
@rpetrusha Thanks for reviewing the exceptions!

rpetrusha commented:
Thanks for making the changes, @tompratt-AQ. This looks good. We should have someone from the PU review this before we merge, though. I'll try to find out who owns it.

rpetrusha commented:
@tompratt-AQ, it turns out that this and the other 2 undocumented MEF namespaces (System.Composition.Hosting and System.Composition.Hosting.Core) were documented in CAPS as OOB releases and, because they weren't part of MREF, weren't migrated to GitHub. Please take a look at them to make sure that we've provided as much detail as the old documentation. The documentation is accessible on MSDN from the [System.Composition namespaces](https://msdn.microsoft.com/en-us/library/windows/apps/jj126278.aspx) topic.

tompratt-AQ commented:
@rpetrusha There is a 4th namespace involved as well, that I just started working on: System.Composition
I will check all of them against MSDN. 
**Question:**  Should I change all documentation for these to the MSDN wording? Or just evaluate the two, modify when they don't match, and add any missing information?

rpetrusha commented:
@tompratt-AQ, evaluate when they don't match and add missing information. They don't have to be identical; select the version that makes the most sense.

tompratt-AQ committed:
Evaluated and made changes based on MSDN documentation.

tompratt-AQ commented:
@rpetrusha I checked the documentation against the [MSDN topics](https://msdn.microsoft.com/en-us/library/windows/apps/jj126278.aspx) and made adjustments accordingly.

tompratt-AQ committed:
Fix tagging issue.

mairaw commented:
@safern @AlexGhiondea do you own this API area? Can you guys help review it?
This content is available to review on staging at https://review.docs.microsoft.com/en-us/dotnet/api/system.composition.convention?branch=pr-en-us-2309&view=netcore-2.0

/cc @karelz 

Thanks!

mairaw commented:
Sorry about that @safern. I read the wrong line at https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/issue-guide.md#areas 😄

rpetrusha commented:
@tompratt-AQ, since this largely duplicates content from MSDN with some improvements, I think it's safe to merge.

AlexGhiondea reviewed (commented):
Thanks for doing this @tompratt-AQ !
